### ISSUE TITLE
error: "object*" is not iterable with zip

### ISSUE THREAD

**Issue by panosz** [microsoft/pyright#5182]
**Describe the bug**
There's an issue with type inference in Pyright when iterating over a set or list containing instances of different classes, even if they share the same attribute structure. The problem arises when these attributes are passed to the zip function and then unpacked. Notably, this issue doesn't occur when the collection is a tuple.

**To Reproduce**

Given the following code:
```python
class A():
    def __init__(self):
        self.x = [2, 3, 4]


class B():
    def __init__(self):
        self.x = [2, 3, 4]


a = A()
b = B()

for s in (a, b):
    print(s.x)  # OK
    for xi, yi in zip(s.x, s.x):  # OK
        print(xi, yi)

for s in {a, b}:
    print(s.x)  # OK
    for xi, yi in zip(s.x, s.x):  #  error: "object*" is not iterable
        print(xi, yi)

for s in [a, b]:
    print(s.x)  # OK
    for xi, yi in zip(s.x, s.x):  #  error: "object*" is not iterable
        print(xi, yi)

```
we get the errors indicated in the comments

**Expected behavior**
No errors should be reported

**VS Code extension or command-line**
command line version 1.1.309

**Comment by erictraut** [comment]
This was fixed in 1.1.310. If you update, the problem will go away.
